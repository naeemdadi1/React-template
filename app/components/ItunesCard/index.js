import React, { useRef } from 'react';
import PropTypes from 'prop-types';
import { Card, Tooltip, Button } from 'antd';
import { LinkOutlined } from '@ant-design/icons';
import { truncate } from 'lodash';
import styled from 'styled-components';
import { T } from '@components/T';
import If from '../If/index';

const CustomCard = styled(Card)`
  && {
    max-width: 22rem;
    margin: 1rem 0;
  }
`;

const CustomAudio = styled.audio`
  margin: 1rem 0;
`;

const ItunesCard = ({ onClickAction, itune, ituneName }) => {
  const audioRef = useRef();

  const {
    trackName,
    artistViewUrl,
    trackViewUrl,
    collectionName,
    collectionPrice,
    collectionViewUrl,
    artistName,
    previewUrl,
    country,
    currency
  } = itune;

  return (
    <CustomCard
      data-testid="itune-card"
      hoverable
      title={trackName ? trackName : ituneName}
      actions={[
        <Tooltip key="artistViewUrl" title="View Artist">
          <Button type="link" target="_blank" href={artistViewUrl} icon={<LinkOutlined />}></Button>
        </Tooltip>,
        <Tooltip key="trackViewUrl" title="View Track">
          <Button type="link" target="_blank" href={trackViewUrl} icon={<LinkOutlined />}></Button>
        </Tooltip>,
        <Tooltip key="collectionViewUrl" title="View Collection">
          <Button type="link" target="_blank" href={collectionViewUrl} icon={<LinkOutlined />}></Button>
        </Tooltip>
      ]}
    >
      <If condition={artistName}>
        <T id="artist_name" values={{ artistName }} />
      </If>
      <If condition={(collectionPrice, currency)}>
        <T id="collection_price" values={{ collectionPrice, currency }} />
      </If>
      <If condition={country}>
        <T id="country" values={{ country }} />
      </If>
      <If condition={collectionName}>
        <T id="collection_name" values={{ collectionName: truncate(collectionName, { length: 20 }) }} />
      </If>
      <CustomAudio controls data-testid="play_event" onPlay={(e) => onClickAction(e, audioRef)} ref={audioRef}>
        <source src={previewUrl} type="audio/mpeg" />
      </CustomAudio>
    </CustomCard>
  );
};

ItunesCard.propTypes = {
  onClickAction: PropTypes.func,
  currTrack: PropTypes.object,
  itune: PropTypes.object.isRequired,
  ituneName: PropTypes.string
};

export default ItunesCard;
