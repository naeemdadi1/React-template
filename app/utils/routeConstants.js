export default {
  repos: {
    route: '/',
    props: {
      maxwidth: 500,
      padding: 20
    },
    exact: true
  },
  ituneDetail: {
    route: '/track/:id',
    exact: true
  }
};
