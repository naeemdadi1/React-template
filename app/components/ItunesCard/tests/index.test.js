/**
 *
 * Tests for ItunesCard
 *
 */

import { renderWithIntl } from '@app/utils/testUtils';
import React from 'react';
import ItunesCard from '../index';
import { fireEvent } from '@testing-library/dom';
import { itune } from './mockData';

describe('<ItunesCard />', () => {
  let onClickAction, ituneName;

  beforeEach(() => {
    itune, (onClickAction = jest.fn());
    ituneName = 'adele';
  });
  it('should render and match the snapshot', () => {
    const { baseElement } = renderWithIntl(
      <ItunesCard itune={itune} ituneName={ituneName} onClickAction={onClickAction} />
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('should contain 1 ItunesCard component', () => {
    const { getAllByTestId } = renderWithIntl(
      <ItunesCard itune={itune} ituneName={ituneName} onClickAction={onClickAction} />
    );
    expect(getAllByTestId('itune-card').length).toBe(1);
  });

  it('should trigger the onClickAction on play button click', () => {
    const { getByTestId } = renderWithIntl(
      <ItunesCard itune={itune} ituneName={ituneName} onClickAction={onClickAction} />
    );
    fireEvent.play(getByTestId('play_event'));
    expect(onClickAction).toBeCalled();
  });

  it('should render ituneName if trackName is not there', () => {
    itune.trackName = null;
    const { getByTestId } = renderWithIntl(<ItunesCard itune={itune} ituneName={ituneName} />);
    const titleElem = getByTestId('itune-card');
    expect(titleElem).toHaveTextContent(ituneName);
  });
});
