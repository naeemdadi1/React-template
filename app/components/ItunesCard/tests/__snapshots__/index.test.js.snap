// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ItunesCard /> should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="ant-card ant-card-bordered ant-card-hoverable ItunesCard__CustomCard-sijrgi-0 kYhVSA"
      data-testid="itune-card"
    >
      <div
        class="ant-card-head"
      >
        <div
          class="ant-card-head-wrapper"
        >
          <div
            class="ant-card-head-title"
          >
            Doin' What She Likes
          </div>
        </div>
      </div>
      <div
        class="ant-card-body"
      >
        <p
          class="T__StyledText-gjlic1-0 kYcyRq"
          data-testid="t"
        >
          Artist Name: Blake Shelton
        </p>
        <p
          class="T__StyledText-gjlic1-0 kYcyRq"
          data-testid="t"
        >
          Collection Price: 10.99 USD
        </p>
        <p
          class="T__StyledText-gjlic1-0 kYcyRq"
          data-testid="t"
        >
          Country: USA
        </p>
        <p
          class="T__StyledText-gjlic1-0 kYcyRq"
          data-testid="t"
        >
          Collection Name: Based On a True S...
        </p>
        <audio
          class="ItunesCard__CustomAudio-sijrgi-1 hTstUt"
          controls=""
          data-testid="play_event"
        >
          <source
            src="https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview115/v4/98/70/8c/98708cde-2365-38e9-7eab-76786d3ef1a9/mzaf_12146965378941562534.plus.aac.p.m4a"
            type="audio/mpeg"
          />
        </audio>
      </div>
      <ul
        class="ant-card-actions"
      >
        <li
          style="width: 33.333333333333336%;"
        >
          <span>
            <a
              class="ant-btn ant-btn-link ant-btn-icon-only"
              href="https://music.apple.com/us/artist/blake-shelton/189204?uo=4"
              target="_blank"
            >
              <span
                aria-label="link"
                class="anticon anticon-link"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="link"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M574 665.4a8.03 8.03 0 00-11.3 0L446.5 781.6c-53.8 53.8-144.6 59.5-204 0-59.5-59.5-53.8-150.2 0-204l116.2-116.2c3.1-3.1 3.1-8.2 0-11.3l-39.8-39.8a8.03 8.03 0 00-11.3 0L191.4 526.5c-84.6 84.6-84.6 221.5 0 306s221.5 84.6 306 0l116.2-116.2c3.1-3.1 3.1-8.2 0-11.3L574 665.4zm258.6-474c-84.6-84.6-221.5-84.6-306 0L410.3 307.6a8.03 8.03 0 000 11.3l39.7 39.7c3.1 3.1 8.2 3.1 11.3 0l116.2-116.2c53.8-53.8 144.6-59.5 204 0 59.5 59.5 53.8 150.2 0 204L665.3 562.6a8.03 8.03 0 000 11.3l39.8 39.8c3.1 3.1 8.2 3.1 11.3 0l116.2-116.2c84.5-84.6 84.5-221.5 0-306.1zM610.1 372.3a8.03 8.03 0 00-11.3 0L372.3 598.7a8.03 8.03 0 000 11.3l39.6 39.6c3.1 3.1 8.2 3.1 11.3 0l226.4-226.4c3.1-3.1 3.1-8.2 0-11.3l-39.5-39.6z"
                  />
                </svg>
              </span>
            </a>
          </span>
        </li>
        <li
          style="width: 33.333333333333336%;"
        >
          <span>
            <a
              class="ant-btn ant-btn-link ant-btn-icon-only"
              href="https://music.apple.com/us/album/doin-what-she-likes/606057263?i=606057326&uo=4"
              target="_blank"
            >
              <span
                aria-label="link"
                class="anticon anticon-link"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="link"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M574 665.4a8.03 8.03 0 00-11.3 0L446.5 781.6c-53.8 53.8-144.6 59.5-204 0-59.5-59.5-53.8-150.2 0-204l116.2-116.2c3.1-3.1 3.1-8.2 0-11.3l-39.8-39.8a8.03 8.03 0 00-11.3 0L191.4 526.5c-84.6 84.6-84.6 221.5 0 306s221.5 84.6 306 0l116.2-116.2c3.1-3.1 3.1-8.2 0-11.3L574 665.4zm258.6-474c-84.6-84.6-221.5-84.6-306 0L410.3 307.6a8.03 8.03 0 000 11.3l39.7 39.7c3.1 3.1 8.2 3.1 11.3 0l116.2-116.2c53.8-53.8 144.6-59.5 204 0 59.5 59.5 53.8 150.2 0 204L665.3 562.6a8.03 8.03 0 000 11.3l39.8 39.8c3.1 3.1 8.2 3.1 11.3 0l116.2-116.2c84.5-84.6 84.5-221.5 0-306.1zM610.1 372.3a8.03 8.03 0 00-11.3 0L372.3 598.7a8.03 8.03 0 000 11.3l39.6 39.6c3.1 3.1 8.2 3.1 11.3 0l226.4-226.4c3.1-3.1 3.1-8.2 0-11.3l-39.5-39.6z"
                  />
                </svg>
              </span>
            </a>
          </span>
        </li>
        <li
          style="width: 33.333333333333336%;"
        >
          <span>
            <a
              class="ant-btn ant-btn-link ant-btn-icon-only"
              href="https://music.apple.com/us/album/doin-what-she-likes/606057263?i=606057326&uo=4"
              target="_blank"
            >
              <span
                aria-label="link"
                class="anticon anticon-link"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="link"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M574 665.4a8.03 8.03 0 00-11.3 0L446.5 781.6c-53.8 53.8-144.6 59.5-204 0-59.5-59.5-53.8-150.2 0-204l116.2-116.2c3.1-3.1 3.1-8.2 0-11.3l-39.8-39.8a8.03 8.03 0 00-11.3 0L191.4 526.5c-84.6 84.6-84.6 221.5 0 306s221.5 84.6 306 0l116.2-116.2c3.1-3.1 3.1-8.2 0-11.3L574 665.4zm258.6-474c-84.6-84.6-221.5-84.6-306 0L410.3 307.6a8.03 8.03 0 000 11.3l39.7 39.7c3.1 3.1 8.2 3.1 11.3 0l116.2-116.2c53.8-53.8 144.6-59.5 204 0 59.5 59.5 53.8 150.2 0 204L665.3 562.6a8.03 8.03 0 000 11.3l39.8 39.8c3.1 3.1 8.2 3.1 11.3 0l116.2-116.2c84.5-84.6 84.5-221.5 0-306.1zM610.1 372.3a8.03 8.03 0 00-11.3 0L372.3 598.7a8.03 8.03 0 000 11.3l39.6 39.6c3.1 3.1 8.2 3.1 11.3 0l226.4-226.4c3.1-3.1 3.1-8.2 0-11.3l-39.5-39.6z"
                  />
                </svg>
              </span>
            </a>
          </span>
        </li>
      </ul>
    </div>
  </div>
</body>
`;
